src/progress.rs

```diff
@@ -47,118 +47,187 @@ impl Default for ProgressState {
+            progress: HashMap::with_capacity_and_hasher(
+                voters + learners,
+                FxBuildHasher::default(),
+            ),
+            configuration: Configuration {
+                voters: HashSet::with_capacity_and_hasher(voters, FxBuildHasher::default()),
+                learners: HashSet::with_capacity_and_hasher(learners, FxBuildHasher::default()),
+            },
         }
     }
 
     /// Returns the status of voters.
-    pub fn voters(&self) -> &FxHashMap<u64, Progress> {
-        &self.voters
+    #[inline]
+    pub fn voters(&self) -> impl Iterator<Item = (&u64, &Progress)> {
+        self.progress
+            .iter()
+            .filter(move |(&k, _)| self.has_voter(k))
     }
 
     /// Returns the status of learners.
-    pub fn learners(&self) -> &FxHashMap<u64, Progress> {
-        &self.learners
+    #[inline]
+    pub fn learners(&self) -> impl Iterator<Item = (&u64, &Progress)> {
+        self.progress
+            .iter()
+            .filter(move |(&k, _)| self.has_learner(k))
+    }
+
+    /// Returns the mutable status of voters.
+    #[inline]
+    pub fn voters_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
+        let set = self.voter_ids().cloned().collect::<FxHashSet<_>>();
+        self.progress
+            .iter_mut()
+            .filter(move |(&k, _)| set.contains(&k))
+    }
+
+    /// Returns the mutable status of learners.
+    #[inline]
+    pub fn learners_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
+        let set = self.learner_ids().cloned().collect::<FxHashSet<_>>();
+        self.progress
+            .iter_mut()
+            .filter(move |(&k, _)| set.contains(&k))
+    }
+
+    /// Returns if the progress set contains a voter by the given id.
+    #[inline]
+    pub fn has_voter(&self, id: u64) -> bool {
+        self.configuration.voters.contains(&id)
+    }
+
+    /// Returns the number of voters in the current configuration.
+    #[inline]
+    pub fn num_voters(&self) -> usize {
+        self.configuration.voters.len()
+    }
+
+    /// Returns the number of learners in the current configuration.
+    #[inline]
+    pub fn num_learners(&self) -> usize {
+        self.configuration.learners.len()
+    }
+
+    /// Returns if the progress set contains a learner by the given id.
+    #[inline]
+    pub fn has_learner(&self, id: u64) -> bool {
+        self.configuration.learners.contains(&id)
     }
 
-    /// Returns the ids of all known nodes.
-    pub fn nodes(&self) -> Vec<u64> {
-        let mut nodes = Vec::with_capacity(self.voters.len());
-        nodes.extend(self.voters.keys());
-        nodes.sort();
-        nodes
+    /// Returns if the progress set contains a peer by the given id.
+    #[inline]
+    pub fn has_peer(&self, id: u64) -> bool {
+        self.progress.contains_key(&id)
+    }
+
+    /// Returns the ids of all known voters.
+    #[inline]
+    pub fn voter_ids(&self) -> impl Iterator<Item = &u64> {
```

Comment:
How about returning `&HashSet<u64>`? So `has_xxx` and `num_xxx` won't be necessary.

Reply:
Yeah, after using the API a bit I agree with this.